lib/configurators/databases/aws.go

```diff
@@ -0,0 +1,656 @@
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		_, err = iamClient.PutRolePermissionsBoundary(&iam.PutRolePermissionsBoundaryInput{
+			PermissionsBoundary: aws.String(policyBoundaryArn),
+			RoleName:            aws.String(a.target),
+		})
+		if err != nil {
+			return wrapAWSError(err)
+		}
+	default:
+		return trace.Errorf("invalid target type %q", a.targetType.String())
+	}
+
+	return nil
+}
+
+// getAWSIAMClient returns the struct `awsIAMClient` if initialized, otherwise
+// creates a new IAM client.
+func (a *awsConfigurator) getAWSIAMClient() (iamiface.IAMAPI, error) {
+	if a.awsIAMClient != nil {
+		return a.awsIAMClient, nil
+	}
+
+	session, err := a.getAWSSession()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	a.awsIAMClient = iam.New(session)
+	return a.awsIAMClient, nil
+}
+
+// getAWSSTSClient returns the struct `awsSTSClient` if initialized, otherwise
+// creates a new STS client.
+func (a *awsConfigurator) getAWSSTSClient() (stsiface.STSAPI, error) {
+	if a.awsSTSClient != nil {
+		return a.awsSTSClient, nil
+	}
+
+	session, err := a.getAWSSession()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	a.awsSTSClient = sts.New(session)
+	return a.awsSTSClient, nil
+}
+
+// getAWSSession returns the struct `awsSession` if initialized, otherwise
+// creates a new Session.
+func (a *awsConfigurator) getAWSSession() (*awssession.Session, error) {
+	if a.awsSession != nil {
+		return a.awsSession, nil
+	}
+
+	var err error
+	a.awsSession, err = awssession.NewSessionWithOptions(awssession.Options{
+		// TODO: remove this, only used to test locally
+		// Config: aws.Config{
+		// 	Endpoint: aws.String("http://localhost:4566"),
+		// },
+		SharedConfigState: awssession.SharedConfigEnable,
+	})
+
+	return a.awsSession, err
+}
+
+// policiesTarget defines which target and its type the policies will be
+// attached to.
+func policiesTarget(flags BootstrapFlags, awsIdentity awslib.Identity) (string, targetType, error) {
+	if flags.AttachToUser != "" {
+		return flags.AttachToUser, targetTypeUser, nil
+	}
+
+	if flags.AttachToRole != "" {
+		return flags.AttachToRole, targetTypeRole, nil
+	}
+
+	if awsIdentity == nil {
+		return defaultAttachTarget, defaultAttachTargetType, nil
+	}
+
+	// If either role or user are not provided, try to get the current one (using
+	// Security token service).
+	target := awsIdentity.GetName()
+	var targetType targetType
+
+	switch awsIdentity.(type) {
+	case awslib.User:
+		targetType = targetTypeUser
+	case awslib.Role:
+		targetType = targetTypeRole
+	default:
+		return "", targetTypeUnknown, trace.BadParameter("not able to identify the target type")
+	}
+
+	return target, targetType, nil
+}
+
+// buildPolicyDocuments generates policy and policy boundary document based on
+// flags and fileConfig.
+func buildPolicyDocuments(targetType targetType, flags BootstrapFlags, fileConfig *config.FileConfig, identity awslib.Identity) (*awslib.PolicyDocument, *awslib.PolicyDocument, error) {
```

Comment:
I think there's too much going on in this method, it's pretty hard to follow.

First, I would split it in 2: separate for generating policy document and boundary document. This will align pretty well with my suggestion above about having separate "steps".

Also, I would refactor it so the flow/logic is more clear, for example:

```go
func buildPolicyDocument(...) (*awslib.PolicyDocument, error) {
  var statements []xxx
  if rdsEnabled() {
    rdsStatements, err := getRDSStatements()
    // check err
    statements = append(statements, rdsStatements...)
  }
  if auroraEnabled() {
    // getAuroraStatements()
  }
  if redshiftEnabled() {
    // getRedshiftStatements()
  }
  return awslib.NewPolicyDocument(statements)
}

func buildBoundaryDocument(...) (*awslib.PolicyDocument, error) {
  // similar to above
}
```